File: docker/ruby/srcs/ft_transcendence/app/javascript/channels/notifications_channel.js
import consumer from "./consumer"
import matchmaking from "./matchmaking_channel"

var timer_queue;
var interval_matchmaking;

var add_notif = function(text, new_id)
{
	let new_one = $(".alert").last().clone();
	new_one.attr("id", new_id);
	$("#alerts-div").append(new_one);
	new_one.find("#alert-text").html(text);
	return (new_one);	
}

var clear_matchmaking = function()
{
	matchmaking.perform("unsubscribe_queue");
	clearInterval(interval_matchmaking);
};

var notif = consumer.subscriptions.create("NotificationsChannel", {

	connected() {
	  	console.log("connected to notif");
	},

	disconnected() {
	},

	received(data) {
		console.log(data);
		if (data.type && data.type == "redirect")
			Turbolinks.visit("/game/" + data.game.id);
		else if (data.type && data.type == "invitation")
			add_notif(data.message + " <a href='/game/" + data.game.id + "'>Click to join</a>").show();
		else if (data.type && data.type == "in_queue")
		{
			if (!$("#matchmaking-alert").length)
			{
				var text = data.message + " - " + "<span id='time_queue'>00:00</span>";
				var new_notif = add_notif(text, "matchmaking-alert");
				var timer_start = Date.now();
				interval_matchmaking = setInterval(function() {
					console.log("interval");
					var res = (Date.now() - timer_start) / 1000;
					var minutes = Math.floor(res / 60) % 60;
					var seconds = Math.floor(res % 60);
					if (minutes < 10)
						minutes = "0" + minutes;
					if (seconds < 10)
						seconds = "0" + seconds;
					$("#time_queue").html(minutes+":"+seconds);
				}, 1000);
				new_notif.find(".close").click(clear_matchmaking);
				new_notif.show();
			}
		}
		else
		{
			createToast("New notification: " + data.title + "<br><a href=" + data.href + ">Open</a>");
		}
	}
});

document.addEventListener('turbolinks:load', () => {
	var button = $("#game_invite");
	var default_button = $("#default");
	if (button)
	{
		button.click(function() {
			console.log("clicked");
			var canvas_width = $("#cwidth").val();
			var canvas_height = $("#cheight").val();
			var ball_radius = $("#bradius").val();
			var max_points = $("#max_points").val();
			var to_nickname = $("#nickname").html();
			notif.perform('send_notif', {
				type: "play_casual",
				to: to_nickname,
				canvas: {
					width: canvas_width,
					height: canvas_height
				},
				ball: {
					radius: ball_radius
				},
				max_points: max_points
			});
		});
	}
	if (default_button)
	{
		default_button.click(function() {
			var canvas_width = $("#cwidth").val(600);
			var canvas_height = $("#cheight").val(600);
			var ball_radius = $("#bradius").val(10);
			var max_points = $("#max_points").val(5);
		});
	}
});

export { interval_matchmaking, notif };

File: docker/ruby/srcs/ft_transcendence/app/channels/notifications_channel.rb
class NotificationsChannel < ApplicationCable::Channel
	include Rails.application.routes.url_helpers
	include ActionView::Helpers::UrlHelper

	def subscribed
		current_user.reload
		stream_for current_user
		# stream_from "notifications_channel"
	end

	def send_notif(data)
		valfrom_user = current_user
		if data["type"] == "play_casual"
			if GameLogic.check_rules(data)
				valto_user = User.find_by(nickname: data["to"])
				if valfrom_user != valto_user && valto_user
					valgame_rules = GameRule.create(canvas_width: data["canvas"]["width"].to_i, canvas_height: data["canvas"]["height"].to_i, ball_radius: data["ball"]["radius"].to_i, max_points: data["max_points"].to_i)
					valgame = Game.create(player1: valfrom_user, player2: valto_user, status: "waiting", mode: "casual", game_rules: valgame_rules)
					valmessage = valfrom_user.nickname + " invited you to play."
					NotificationsChannel.broadcast_to(valto_user, {
						type: "invitation",
						game:
						{
							id: valgame.id
						},
						message: valmessage
					})
#					Notification.create(user: valto_user, title: 'Invitation to play', message: "#{<%= link_to valfrom_user.nickname, profile_path(valfrom_user)} invited you to play a game!<br>Join #{link_to 'here', game_path(valgame)}")
					NotificationsChannel.broadcast_to(valfrom_user, {
						type: "redirect",
						game:
						{
							id: valgame.id
						}
					})
				end
			end
		end
		if data["type"] == "play_matchmaking"
			valmessage = "Waiting for matchmaking"
			NotificationsChannel.broadcast_to(valfrom_user, {
				type: "in_queue",
				message: valmessage
			});
		end
	end

	def unsubscribed
	# Any cleanup needed when channel is unsubscribed
	end

end
